theano/tensor/opt.py

```diff
@@ -1921,6 +1921,34 @@ def local_set_to_inc_subtensor(node):
 
 @register_canonicalize
 @register_specialize
+@gof.local_optimizer([Subtensor])
+def local_useless_slice(node):
+    """
+    Remove Subtensor of the form X[0, :] -> X[0]
+    """
+    if isinstance(node.op, Subtensor):
+        slices = get_idx_list(node.inputs, node.op.idx_list)
+        last_slice = len(slices)
+        for s in slices[::-1]: 
+            # check if slice and then check slice indices
+            if isinstance(s, slice): 
+                if (s.start is None and s.stop is None and
+                   (s.step is None or T.extract_constant(s.step) == 1)): 
```

Comment:
I think I know why some tests fail.  You should merge the conditions from these two ifs otherwise, you can get something which is a slice, but does not pass the second test, which will continue the loop until you get a non-slice.  It will count intermediate slices which are full of `None` as to be removed.